### PR TITLE
MAGN-8735 Re-enable Visualization tests and move into shared assembly. Update job on CI to target new visualization assembly.

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

During the recent addition of an updated `HelixWatch3DViewModel` component, it was found that, on our CI systems, we could not create a rendering context during system testing due to the lack of DirectX 10 features on the test machines. A temporary solution was enacted to shut off the creation of the background preview when running tests. But for a number of tests, a proper rendering context is required. These tests were marked as failures in order to get the build running again.

This PR replaces the temporary solution, checking for `DynamoModel.IsTestMode` and conditionally creating the view, with a more robust solution where the type of `Watch3DViewModel` is passed into the constructor of the `DynamoViewModel` through the `StartupParams`. By default, a `DefaultWatch3DViewModel` will be created which allows for the running of system tests, but does not actually render anything. For sandbox, Revit, and the visualization tests, we inject a `HelixWatch3DViewModel` which does render.

A new static constructor has been added,`HelixWatch3DViewModel.TryCreateHelixWatch3DViewModel(...)`. This method of construction attempts to create a valid `HelixWatch3DViewModel`, and if it cannot be created due to the lack of resources on the client, a `DefaultWatch3DViewModel` is created, a message is logged to the console, and the exception is written to the log.

All tests requiring a full 3d rendering context are moved to a new assembly, `WpfVisualizationTests`, where a new base class `VisualizationTest` constructs the `DynamoViewModel` using a `HelixWatch3DViewModel` and thus establishing a proper rendering context.

To test this functionality, we can use the DirectX debugging tools. When we set the "Feature Level limit" to 9_3, and we start Dynamo.
image
Dynamo starts with the background preview disabled. The background preview selection is disabledin the View menu as well.
image
Turning off the feature level limit, and restarting Dynamo puts us, once again, into the mode where the background preview is disabled, as this was the last saved state. Enabling the background preview in the view model returns us to 3D:
image

### Related PRs:
https://github.com/DynamoDS/DynamoRevit/pull/681

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@mjkkirschner 

### FYIs
@aparajit-pratap These changes affect the work you are doing in https://github.com/DynamoDS/Dynamo/pull/5418
@limrzx - The existing visualization tests on the CI, will now need to run tests from the `WpfVisualizationTests.dll`

### CODE REVIEW

File: src/VisualizationTests/HelixWatch3DViewModelTests.cs [hunk 1]
```diff
@@ -38,6 +47,55 @@ protected override void GetLibrariesToPreload(List<string> libraries)
...
+            //create the view
+            View = new DynamoView(ViewModel);
+            View.Show();
+
+            SynchronizationContext.SetSynchronizationContext(new SynchronizationContext());
+        }
+    }
+
+    // TODO: http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8735
+    [TestFixture, Category("Failure")]
```
Comment (mjkkirschner) [inline src/VisualizationTests/HelixWatch3DViewModelTests.cs]: should still be marked failure and TODO?

File: test/DynamoCoreWpfTests/ConverterTests.cs [hunk 2]
```diff
@@ -20,12 +20,6 @@ namespace Dynamo.Tests
 {
     class ConverterTests
     {
-        [SetUp]
-        public void Setup()
-        {
-            DynamoModel.IsTestMode = true;
```
Comment (mjkkirschner) [inline test/DynamoCoreWpfTests/ConverterTests.cs]: is this an unrelated cleanup?

**Comment by mjkkirschner** [comment]
@ikeough it looks like there are some conflicts now, I've looked through the changes here and will need a another look to understand them all, but can you fix the conflicts before I take another look?

File: src/Dynamo.All.2013.sln [hunk 3]
```diff
@@ -209,6 +209,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DynamoPackages", "DynamoPac
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "PythonNodeModels", "Libraries\PythonNodeModels\PythonNodeModels.csproj", "{8872CA17-C10D-43B9-8393-5C5A57065EB0}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WpfVisualizationTests", "VisualizationTests\WpfVisualizationTests.csproj", "{C4964946-B367-44EE-9ED2-451FF2A83D32}"
```
Comment (mjkkirschner) [inline src/Dynamo.All.2013.sln]: can you also make these changes to the 2012 .sln since at this time it is still around.... 

I know these are wpf changes - but is there anything that will need to be updated in the mono solution?

File: test/DynamoCoreWpfTests/ConverterTests.cs (hunk 2 above)
Comment (ikeough) [inline test/DynamoCoreWpfTests/ConverterTests.cs]: This is not unrelated. This was put in to ensure that all tests ran `IsTestMode` because that flag was dictating whether the view was created. That flag is no longer dictating whether the view is created, so it is removed.

File: src/Dynamo.All.2013.sln (hunk 3 above)
Comment (ikeough) [inline src/Dynamo.All.2013.sln]: Nothing should require update in the mono solution. But I'll double check.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 4]
```diff
@@ -49,6 +49,8 @@ namespace Dynamo.Controls
     /// </summary>
     public partial class DynamoView : Window, IDisposable
     {
+        public const string BackgroundPreviewName = "BackgroundPreview";
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: should this be pointing to some resource?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 5]
```diff
@@ -375,7 +375,7 @@ public bool EnablePresetOptions
             get { return this.Model.CurrentWorkspace.Presets.Any(); }            
         }
 
-        public List<Watch3DViewModelBase> Watch3DViewModels
+        public List<DefaultWatch3DViewModel> Watch3DViewModels
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: should this have a comment since it's public?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 6]
```diff
@@ -385,7 +385,7 @@ protected set
             }
         }
 
-        public Watch3DViewModelBase BackgroundPreviewViewModel { get; private set; }
+        public DefaultWatch3DViewModel BackgroundPreviewViewModel { get; private set; }
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: same here - public property should have a comment?

File: src/DynamoCore/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -21,4 +21,5 @@
 [assembly: InternalsVisibleTo("DynamoCLI")]
 [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")] // Dynamic assembly created by Moq
 [assembly: InternalsVisibleTo("DynamoStudio")]
-[assembly: InternalsVisibleTo("Reach")]
\ No newline at end of file
+[assembly: InternalsVisibleTo("WpfVisualizationTests")]
```
Comment (mjkkirschner) [inline src/DynamoCore/Properties/AssemblyInfo.cs]: should this new project have an assembly shared info.cs?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 6 above)
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Yes.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 8]
```diff
@@ -501,6 +493,7 @@ private void RenderPackageFactoryViewModel_PropertyChanged(object sender, Proper
                 case "ShowEdges":
                     var factoryVm = (RenderPackageFactoryViewModel)sender;
                     model.PreferenceSettings.ShowEdges = factoryVm.Factory.TessellationParameters.ShowEdges;
+                    Watch3DViewModels.ForEach(vm=>vm.RegenerateAllPackages());
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: can you add a comment as to why showing edges requires regenerating all packages?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 4 above)
Comment (ikeough) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: No. This is not visible in the UI.

File: src/DynamoCore/Properties/AssemblyInfo.cs (hunk 7 above)
Comment (ikeough) [inline src/DynamoCore/Properties/AssemblyInfo.cs]: Yes. Adding it now.

**Comment by ikeough** [comment]
@mjkkirschner The self-service CI is running clean.

File: src/VisualizationTests/HelixWatch3DViewModelTests.cs [hunk 9]
```diff
@@ -45,16 +102,10 @@ private IEnumerable<Model3D> BackgroundPreviewGeometry
...
-
-        [SetUp]
-        public void Setup()
-        {
-            var renderingTier = (RenderCapability.Tier >> 16);
-            if (renderingTier < 2)
+            get
             {
-                Assert.Inconclusive("A watch 3d view cannot be created for this test. Ensure that the testing system is DirectX 11 capable.");
+                return (Watch3DView) View.background_grid.Children().
```
Comment (mjkkirschner) [inline src/VisualizationTests/HelixWatch3DViewModelTests.cs]: could view potentially be null?

**Comment by mjkkirschner** [comment]
@ikeough this looks good to me besides updating the 2012 .sln and one potential null check
